### PR TITLE
Adding support for segments in Period and Representation.

### PR DESCRIPTION
**Pull request by ishita12**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -86,6 +87,7 @@
     "version": "5.0.0"
   },
   "dependencies": {
+    "deepmerge": "^2.0.1",
```
Comment (forbesjo) [inline package.json]: I think we should update the `shallowMerge` function to be a `deepMerge` and remove this dependency

File: src/toPlaylists.js [hunk 2]
```diff
@@ -2,27 +2,25 @@ import { shallowMerge } from './utils/object';
 import { segmentsFromTemplate } from './segment/segmentTemplate';
 import { segmentsFromList } from './segment/segmentList';
 import { segmentsFromBase } from './segment/segmentBase';
-
+// import merge from 'deepmerge';
```
Comment (forbesjo) [inline src/toPlaylists.js]: leftover comment

File: src/inheritAttributes.js [hunk 3]
```diff
@@ -131,14 +132,16 @@ export const getSegmentInformation = (adaptationSet) => {
...
   const repBaseUrlElements = findChildren(representation, 'BaseURL');
   const repBaseUrls = buildBaseUrls(adaptationSetBaseUrls, repBaseUrlElements);
   const attributes = shallowMerge(adaptationSetAttributes, getAttributes(representation));
+  const segmentInfoFromRepresenation = getSegmentInformation(representation);
+  const segmentRepresentationInfo = merge(periodAdaptationSetInfo, segmentInfoFromRepresenation);
 
   return repBaseUrls.map(baseUrl => {
     return {
-      segmentInfo,
+      segmentInfo: segmentRepresentationInfo,
```
[the remark this answers is not in the document]
Comment (ishita12) [reply src/inheritAttributes.js]: Sure I will make that change

File: src/toPlaylists.js [hunk 4]
```diff
@@ -2,27 +2,24 @@ import { shallowMerge } from './utils/object';
 import { segmentsFromTemplate } from './segment/segmentTemplate';
 import { segmentsFromList } from './segment/segmentList';
 import { segmentsFromBase } from './segment/segmentBase';
-
 export const generateSegments = (segmentInfo, attributes) => {
-  if (segmentInfo.template) {
-    return segmentsFromTemplate(
-      shallowMerge(segmentInfo.template, attributes),
-      segmentInfo.timeline
+  if (segmentInfo.list) {
```
Comment (mjneil) [inline src/toPlaylists.js]: Was there a specific reasoning for switching the order of `list` and `template` here? I think either order should be fine, just curious.

File: test/inheritAttributes.test.js [hunk 5]
```diff
@@ -197,6 +198,7 @@ QUnit.test('end to end - basic', function(assert) {
             height="404"
             id="test"
             width="720">
+            <SegmentTemplate></SegmentTemplate>
```
Comment (mjneil) [inline test/inheritAttributes.test.js]: Why was this added?

File: test/inheritAttributes.test.js [hunk 6]
```diff
@@ -272,6 +274,7 @@ QUnit.test('end to end - inherits BaseURL from all levels', function(assert) {
...
```
Comment (mjneil) [inline test/inheritAttributes.test.js]: Why was this one added as well?

File: test/inheritAttributes.test.js [hunk 7]
```diff
@@ -490,3 +494,869 @@ QUnit.test('end to end - alternate BaseURLs', function(assert) {
...
+      periodIndex: 0,
+      role: {
+        value: 'main'
+      },
+      segmentAlignment: 'true',
+      sourceDuration: 30,
+      startWithSAP: '1'
+
+    },
+    segments: undefined
```
Comment (mjneil) [inline test/inheritAttributes.test.js]: This should not be undefined, it should have a segment list built from the segment information provided in the adaptation set

File: test/inheritAttributes.test.js [hunk 8]
```diff
@@ -490,3 +494,869 @@ QUnit.test('end to end - alternate BaseURLs', function(assert) {
...
+      mediaPresentationDuration: 'PT30S',
+      mimeType: 'text/vtt',
+      periodIndex: 0,
+      role: {},
+      sourceDuration: 30
+    },
+    segmentInfo: {
+      base: undefined,
+      list: undefined,
+      template: undefined,
```
Comment (mjneil) [inline test/inheritAttributes.test.js]: This should not be undefined, should have used template from containing adaptation set

File: package.json (hunk 1 above)
Comment (forbesjo) [reply package.json]: When building with `video.js`'s `mergeOptions` it seems like the rollup did not remove the unnecessary code and mpd-parser.js shot up to 680KB. I think it would be better to write a simple deep merge function instead.

File: test/inheritAttributes.test.js [hunk 9]
```diff
@@ -490,3 +494,869 @@ QUnit.test('end to end - alternate BaseURLs', function(assert) {
...
```
Comment (mjneil) [inline test/inheritAttributes.test.js]: This should not be undefined, should have used template from containing adaptation set

File: test/inheritAttributes.test.js [hunk 10]
```diff
@@ -490,3 +494,869 @@ QUnit.test('end to end - alternate BaseURLs', function(assert) {
...
+    }
+  }];
+
+  assert.equal(actual.length, 3);
+  assert.deepEqual(actual, expected);
+});
+
+// Test for only periodAdaptationSetInfo
+
+QUnit.test(' End to End test for returning period information', function(assert) {
```
Comment (mjneil) [inline test/inheritAttributes.test.js]: This test is nearly identical to the one above it . Is it testing something specifically different?

File: test/inheritAttributes.test.js [hunk 11]
```diff
@@ -490,3 +494,869 @@ QUnit.test('end to end - alternate BaseURLs', function(assert) {
...
+           <Initialization range= "0-8888"  />
+          </SegmentBase>
+          <Representation
+            mimeType= "video/mp6"
+            bandwidth= "5000000"
+            codecs= "avc1.64001e"
+            height= "404"
+            id= "test"
+            width= "720" >
+            <SegmentBase></SegmentBase>
```
Comment (mjneil) [inline test/inheritAttributes.test.js]: We should remove this line

File: test/inheritAttributes.test.js [hunk 12]
```diff
@@ -490,3 +494,869 @@ QUnit.test('end to end - alternate BaseURLs', function(assert) {
...
+      template: undefined,
+      timeline: undefined
+    }
+  }];
+
+  assert.equal(actual.length, 3);
+  assert.deepEqual(actual, expected);
+});
+
+QUnit.test(' Test for checking presence of atmost one segment at each level', function(assert) {
```
Comment (mjneil) [inline test/inheritAttributes.test.js]: This description is a bit misleading maybe something along the lines of "uses only one set of segment information when multiple are present" and then add a code comment saying that although the spec states that at each level at most one shall be present, we still handle the case to prevent throwing errors

File: test/manifests/maat_vtt_segmentTemplate.mpd [hunk 13]
```diff
@@ -6,21 +6,33 @@
     <AdaptationSet mimeType="audio/mp4" segmentAlignment="true" startWithSAP="1" lang="en">
       <Role schemeIdUri="urn:mpeg:dash:role:2011" value="main"></Role>
       <SegmentTemplate duration="95232" initialization="$RepresentationID$/init.m4f" media="$RepresentationID$/$Number$.m4f" startNumber="0" timescale="48000"></SegmentTemplate>
-      <Representation audioSamplingRate="48000" bandwidth="63000" codecs="mp4a.40.2" id="63000"></Representation>
-      <Representation audioSamplingRate="48000" bandwidth="125000" codecs="mp4a.40.2" id="125000"></Representation>
+      <Representation audioSamplingRate="48000" bandwidth="63000" codecs="mp4a.40.2" id="63000">
+        <SegmentTemplate></SegmentTemplate>
```
Comment (mjneil) [inline test/manifests/maat_vtt_segmentTemplate.mpd]: all these `<segment*>` tags added to these mpd manifests shouldn't be necessary to add

File: src/toPlaylists.js (hunk 4 above)
Comment (ishita12) [reply src/toPlaylists.js]: I was just checking for the SegmentList output.

File: src/inheritAttributes.js [hunk 14]
```diff
@@ -132,16 +132,20 @@ export const getSegmentInformation = (adaptationSet) => {
...
   const attributes = shallowMerge(adaptationSetAttributes, getAttributes(representation));
-  const segmentInfoFromRepresenation = getSegmentInformation(representation);
-  const segmentRepresentationInfo = merge(periodAdaptationSetInfo, segmentInfoFromRepresenation);
+  const representationSegmentInfo = getSegmentInformation(representation);
 
   return repBaseUrls.map(baseUrl => {
     return {
-      segmentInfo: segmentRepresentationInfo,
+      segmentInfo: {
+        list: adaptationSetSegmentInfo.list && representationSegmentInfo.list ? merge(adaptationSetSegmentInfo.list, representationSegmentInfo.list) : adaptationSetSegmentInfo.list && !representationSegmentInfo.list ? adaptationSetSegmentInfo.list : !adaptationSetSegmentInfo.list && representationSegmentInfo.list ? representationSegmentInfo.list : undefined,
```
Comment (mjneil) [inline src/inheritAttributes.js]: This is a nightmare to read. I think its best we try to come at this with a different approach.

The main issue with the merge is that if the second object in the merge has a property that is `undefined`, it overwrites that property with `undefined`, so removing those properties from the object before we even get to merging should solve the issue without needing all this complex logic.

I think that we should return to what you had before this commit of just merging the full segment info objects instead of merging these properties individually.

Then, to fix the issue of a property being `undefined`, in `getSegmentInformation`, before returning the segment info object, save it to a variable and then delete properties that are undefined
```
const segmentInfo = {
  template,
  timeline: ...
  ....
};

Object.keys(segmentInfo).forEach(key => {
  if (!segmentInfo[key]) {
    delete segmentInfo[key];
  }
});

return segmentInfo;
```

File: src/inheritAttributes.js [hunk 15]
```diff
@@ -181,10 +185,15 @@ export const toRepresentations =
                              roleAttributes);
   const segmentInfo = getSegmentInformation(adaptationSet);
   const representations = findChildren(adaptationSet, 'Representation');
-  const periodAdaptationSetInfo = merge(periodInfo, segmentInfo);
+  const adaptationSetSegmentInfo = {
+    list: periodSegmentInfo.list && segmentInfo.list ? merge(periodSegmentInfo.list, segmentInfo.list) : periodSegmentInfo.list && !segmentInfo.list ? periodSegmentInfo.list : !periodSegmentInfo.list && segmentInfo.list ? segmentInfo.list : undefined,
```
Comment (mjneil) [inline src/inheritAttributes.js]: Same strategy as my other comment here

File: test/manifests/maat_vtt_segmentTemplate.mpd (hunk 13 above)
Comment (OshinKaramian) [reply test/manifests/maat_vtt_segmentTemplate.mpd]: If you wanted to test your use case within an actual manifest, you could create a fresh one, and add that and the accompanying `json` to `index.test.js`.

File: src/inheritAttributes.js [hunk 16]
```diff
@@ -1,10 +1,11 @@
 import { flatten } from './utils/list';
-import { shallowMerge, getAttributes } from './utils/object';
+import { getAttributes } from './utils/object';
 import { parseDuration } from './utils/time';
 import { findChildren, getContent } from './utils/xml';
 import resolveUrl from './utils/resolveUrl';
 import errors from './errors';
-import merge from 'deepmerge';
+// import merge from 'deepmerge';
```
Comment (mjneil) [inline src/inheritAttributes.js]: can remove this comment

File: src/inheritAttributes.js [hunk 17]
```diff
@@ -1,10 +1,11 @@
 import { flatten } from './utils/list';
-import { shallowMerge, getAttributes } from './utils/object';
+import { getAttributes } from './utils/object';
```
Comment (mjneil) [inline src/inheritAttributes.js]: you can import `merge` along with `getAttributes` here

File: src/inheritAttributes.js [hunk 18]
```diff
@@ -1,10 +1,11 @@
...
 import { parseDuration } from './utils/time';
 import { findChildren, getContent } from './utils/xml';
 import resolveUrl from './utils/resolveUrl';
 import errors from './errors';
-import merge from 'deepmerge';
+// import merge from 'deepmerge';
+import { merge } from './utils/object';
```
Comment (mjneil) [inline src/inheritAttributes.js]: can remove this line

File: src/toPlaylists.js (hunk 2 above)
Comment (mjneil) [reply src/toPlaylists.js]: dont forget to remove this comment

File: src/utils/object.js [hunk 19]
```diff
@@ -1,13 +1,24 @@
 import { from } from './list';
 
-export const shallowMerge = (...objects) => {
+export const merge = (...objects) => {
+
+  const isObject = (obj) => {
```
Comment (mjneil) [inline src/utils/object.js]: this function can be defined outside of `merge`

File: src/utils/object.js [hunk 20]
```diff
@@ -1,13 +1,24 @@
...
+    return obj && typeof obj === 'object';
```
Comment (mjneil) [inline src/utils/object.js]: lets be explicit and use `!!obj`

File: src/utils/object.js [hunk 21]
```diff
@@ -1,13 +1,24 @@
...
+  };
+
   return objects.reduce((x, y) => {
```
Comment (mjneil) [inline src/utils/object.js]: I think it would be nice to update this to use more descriptive names for readability. maybe `result` for `x` and `source` for `y`

File: test/utils.test.js [hunk 22]
```diff
@@ -8,25 +8,25 @@ import QUnit from 'qunit';
 
 QUnit.module('utils');
 
-QUnit.module('shallowMerge');
+QUnit.module('merge');
```
Comment (forbesjo) [inline test/utils.test.js]: Might want to add a test or two for the new merge functionality
